syntax/asciidoctor.vim

```diff
@@ -53,11 +53,7 @@ syn region asciidoctorH4 start="^=====\s" end="$" oneline contains=@asciidoctorI
 syn region asciidoctorH5 start="^======\s" end="$" oneline contains=@asciidoctorInline,@Spell
 syn region asciidoctorH6 start="^=======\s" end="$" oneline contains=@asciidoctorInline,@Spell
 
-syn match asciidoctorH1 '^\%(\n\|\%^\)\k\+.*\n==\+\n$'   contains=@Spell
-syn match asciidoctorH2 '^\%(\n\|\%^\)\k\+.*\n--\+\n$'   contains=@Spell
-syn match asciidoctorH3 '^\%(\n\|\%^\)\k\+.*\n\~\~\+\n$' contains=@Spell
-syn match asciidoctorH4 '^\%(\n\|\%^\)\k\+.*\n\^\^\+\n$' contains=@Spell
-syn match asciidoctorH5 '^\%(\n\|\%^\)\k\+.*\n++\+\n$'   contains=@Spell
+syn match asciidoctorH '^\%(\n\|\%^\)\k\+.*\n[=\-~^+]\{2,}\n$'   contains=@Spell
```

Comment:
call it `asciidoctorSetextHeader` maybe? And add link to Title (`hi def link ...`)

Reply:
Also it would make 

```
header
=+^--

```

highlighted, right?

Reply:
This fails to highlight:
![image](https://user-images.githubusercontent.com/234774/69937781-ae759400-14ec-11ea-930a-fd1e796e746f.png)

Reply:
Yes, is it a problem? The same restrictions (empty lines before/after) apply, and performance wasn't too good... I'd still prefer it this way, but I didn't think about that.

Reply:
If you add empty line after it highlights it:
![image](https://user-images.githubusercontent.com/234774/69938057-5723f380-14ed-11ea-8486-5afbedf8f5f1.png)

As far as I know asiidoctor doesn't require empty line after setext header

Reply:
I did it to avoid conflicts with other block types, now I removed that requirement.

Reply:
Could you test

```
syn match asciidoctorSetextHeader '^\%(\n\|\%^\)\k\+.*\n\%(=\|\-\|\~\|\^\|+\)\{2,}\n'   contains=@Spell
hi link asciidoctorSetextHeader Title
```

For speed?

Reply:
The unified match is as fast as each of the previous ones... that is 5x faster. I didn't make exact measurements but the ratio with the slowest match is about the same.

Reply:
No, I mean with my changes
```
syn match asciidoctorSetextHeader '^\%(\n\|\%^\)\k\+.*\n\%(=\|\-\|\~\|\^\|+\)\{2,}$'   contains=@Spell
```

Where no mixed underline is allowed.

Is it significantly slower than your `[...]`?

Reply:
It's slightly slower, anyway it doesn't fix the issue with possible mixed characters
```
  TOTALE     CONT.  CORRIS. PIU LENTO   MEDIA     NOME               MODELLO
  0.799160   65461  0       0.001725    0.000012  asciidoctorAdmonition \C^\%(NOTE:\)\|\%(TIP:\)\|\%(IMPORTANT:\)\|\%(CAUTION:\)\|\%(WARNING:\)\s
  0.538491   65461  0       0.001102    0.000008  asciidoctorUrl     \%(link\|xref\|image\)::\?
  0.483982   65461  0       0.000694    0.000007  asciidoctorBold    \m\%(^\|[[:punct:][:space:]]\@<=\)\*\ze[^* ].\{-}\S
  0.476070   65461  0       0.000103    0.000007  asciidoctorBoldItalic \m\%(^\|[[:punct:][:space:]]\@<=\)\*_\ze[^*_ ].\{-}\S
  0.459019   67120  1659    0.000605    0.000007  asciidoctorCode    \m\%(^\|[[:punct:][:space:]]\@<=\)`\ze[^` ].\{-}\S
  0.452812   67866  2405    0.000730    0.000007  asciidoctorItalic  \m\%(^\|[[:punct:][:space:]]\@<=\)_\ze[^_ ].\{-}\S
  0.399532   65461  0       0.000122    0.000006  asciidoctorUrl     \%(http\|ftp\)s\?://\S\+\ze\%(\[.\{-}\]\)
  0.378971   65461  0       0.000873    0.000006  asciidoctorMacro   \a\+::\?\(\S[[:alnum:][:blank:]./\\:-]\{-}\)\?\[.\{-}\]
  0.351361   65461  0       0.000953    0.000005  asciidoctorDefList .\{-}::\_s\%(\_^\n\)\?
  0.189066   65461  6267    0.000081    0.000003  asciidoctorIndented ^+\?\n\%(\s\+\(-\|[*.]\+\|\d\+\.\|\a\.\)\s\)\@!\(\s.*\n\)\+
  0.140309   65461  1454    0.000074    0.000002  asciidoctorSetextHeader ^\%(\n\|\%^\)\k\+.*\n\%(=\|\-\|\~\|\^\|+\)\{2,}$
  0.096571   65504  43      0.000029    0.000001  asciidoctorListMarker ^\s*\(-\|\*\+\|\.\+\)\%(\s\+\[[Xx ]\]\+\s*\)\?\%(\s\+\S\)\@=


  5.841858   3466320
```

Reply:
> anyway it doesn't fix the issue with possible mixed characters

Yes, you're right.

What about
```
syn match asciidoctorSetextHeader '^\%(\n\|\%^\)\k\+.*\n\%(=\{2,}\|\-\{2,}\|\~\{2,}\|\^\{2,}\|+\{2,}\)$'   contains=@Spell
```
Or do I overengineer here? :)

Reply:
I didn't see this comment, the current match only checks for one character. This would also work:

    syn match asciidoctorSetextHeader '^\%(\n\|\%^\)\k.*\n\%(==\+\|\-\-\+\|\~\~\+\|^\^\+\|++\+\)$'   contains=@Spell

But 2+ characters don't seem to be required (1 underlined character is rendered as header in my test) so I'd leave it as it is now.